test/patterns_as_and_drop/test.ml

```diff
@@ -0,0 +1,81 @@
+open Ppxlib
+open Format
+
+let loc = Location.none
+
+let ast = [%expr List.length xs = 0]
+
+(* Simple demo without [as__] and [drop] *)
+let () =
+  let pat =
+    let open Ast_pattern in
+    let length () =
+      pexp_apply
+        (pexp_ident (ldot (lident (string "List")) (string "length")))
+        ((nolabel ** __) ^:: nil)
+    in
+    let zero () = pexp_constant (pconst_integer (string "0") none) in
+
+    pexp_apply
+      (pexp_ident (lident (string "=")))
+      ((nolabel ** length ()) ^:: (nolabel ** zero ()) ^:: nil)
+  in
+
+  Ast_pattern.parse pat loc ast
+    ~on_error:(fun () -> print_endline "error")
+    (fun _ -> print_endline "success")
+
+let () =
+  let pat =
+    let open Ast_pattern in
+    let length () =
+      as__
+        (pexp_apply
+           (pexp_ident (ldot (lident (string "List")) (string "length")))
+           ((nolabel ** drop) ^:: nil))
+    in
+    let zero () = as__ (pexp_constant (pconst_integer (string "0") none)) in
+
+    pexp_apply
+      (pexp_ident (lident (string "=")))
+      ((nolabel ** length ()) ^:: (nolabel ** zero ()) ^:: nil)
+  in
+
+  Ast_pattern.parse pat loc ast
+    ~on_error:(fun () -> print_endline "error")
+    (fun _l _r ->
+      (* Format.printf "'%a' and '%a'\n%!" Pprintast.expr l Pprintast.expr r; *)
+      (* Error: The module Pprintast is an alias for module Astlib.Pprintast, which is missing *)
+      Format.printf "Success but '??' and '??'\n%!")
+
+(** Pitfall. If we forget unit argument and will use [as__], the success case
+    will be fired before the error case. *)
+let () =
+  let loc = Location.none in
+  let ast = [%expr 1] in
+
+  let pat () =
+    let open Ast_pattern in
+    as__ (pexp_constant (pconst_integer (string "0") none))
+  in
+  printf "Wrong!\n%!";
+  Ast_pattern.parse (pat ()) loc ast
+    ~on_error:(fun () -> print_endline "error")
+    (fun _zero_expr -> Format.printf "Success and error after that\n%!");
+  printf "Correct!\n%!";
+  Ast_pattern.parse (pat ()) loc ast
+    ~on_error:(fun () () -> print_endline "error")
+    (fun _zero_expr () -> Format.printf "Success and error after that\n%!")
+    ()
+
+(* But this pitfall is not introduced by [as__], it was before too. *)
+let () =
+  let ast = [%expr string_of_int 42] in
+  let pat =
+    let open Ast_pattern in
+    pexp_apply (pexp_ident (lident __)) ((nolabel ** eint (int 42)) ^:: nil)
+  in
+
+  Ast_pattern.parse pat loc ast
+    ~on_error:(fun () -> print_endline "error")
+    (fun s -> printf "success %s, but actually not" s)
```

Comment:
How does this new test reveal the pitfall? At first sight, it seems like the `ast` you define simply fulfils the pattern you define. So why is this "actually no success"? (I'm probably be missing something)

Reply:
Actually, it was a little bit crash and this test doesn't represent what I wanted. Not it should be better.

Reply:
Thanks!